src/main/java/core/Game.java

```diff
@@ -545,11 +547,18 @@ public final AbstractAction oneAction() {
                 action = observedActions.get(0);
                 currentPlayer.registerUpdatedObservation(observation);
             } else {
-                // Get action from player, and time it
-                s = System.nanoTime();
-                if (debug) System.out.printf("About to get action for player %d%n", gameState.getCurrentPlayer());
-                action = currentPlayer.getAction(observation, observedActions);
-                agentTime += (System.nanoTime() - s);
+                boolean explore = rnd.nextDouble() < exploreEpsilon;
```

Comment:
Why is this done as part of the game? I think it's complicating the game loop and I don't necessarily agree from a structural point of view that this should break the expectations of 'I give an action to the game and the game applies it'. If a different training procedure is needed, then I think this should be set up separately (even if it's probably not as easy...).

Reply:
I agree - this was always terribly ugly.

I've now refactored this into AbstractPlayer which is a much better place for it. It can't really be any lower in the hierarchy as this could be used with any implementation.